flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java

```diff
@@ -0,0 +1,203 @@
+        configuration.set(NettyShuffleEnvironmentOptions.BATCH_SHUFFLE_COMPRESSION_ENABLED, false);
+        return configuration;
+    }
+
+    private JobGraph createJobGraph(
+            int numRecordsToSend, boolean failExecution, Configuration configuration) {
+        StreamExecutionEnvironment env =
+                StreamExecutionEnvironment.getExecutionEnvironment(configuration);
+        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(10, 0L));
+        env.setParallelism(NUM_SLOTS_PER_TASK_MANAGER);
+
+        DataStream<String> source =
+                new DataStreamSource<>(
+                        env,
+                        BasicTypeInfo.STRING_TYPE_INFO,
+                        new StreamSource<>(new StringSource(numRecordsToSend)),
+                        true,
+                        "source",
+                        Boundedness.BOUNDED);
+        source.rebalance().map(value -> value).shuffle().addSink(new VerifySink(failExecution));
+
+        StreamGraph streamGraph = env.getStreamGraph();
+        streamGraph.setJobType(JobType.BATCH);
+        return StreamingJobGraphGenerator.createJobGraph(streamGraph);
+    }
+
+    private void checkAllDataReceived(int numRecordsToSend) {
+        assertThat(Arrays.stream(NUM_RECEIVED_RECORDS).sum())
+                .isEqualTo(numRecordsToSend * PARALLELISM);
+    }
+
+    private static class StringSource implements ParallelSourceFunction<String> {
+        private volatile boolean isRunning = true;
+        private int numRecordsToSend;
+
+        StringSource(int numRecordsToSend) {
+            this.numRecordsToSend = numRecordsToSend;
+        }
+
+        @Override
+        public void run(SourceContext<String> ctx) throws Exception {
+            while (isRunning && numRecordsToSend-- > 0) {
+                ctx.collect(RECORD);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            isRunning = false;
+        }
+    }
+
+    private static class VerifySink extends RichSinkFunction<String> {
+        private final boolean failExecution;
+
+        VerifySink(boolean failExecution) {
+            this.failExecution = failExecution;
+        }
+
+        @Override
+        public void open(Configuration parameters) throws Exception {
+            NUM_RECEIVED_RECORDS[getRuntimeContext().getIndexOfThisSubtask()] = 0;
+            if (!failExecution
+                    || getRuntimeContext().getAttemptNumber() > 0
+                    || getRuntimeContext().getIndexOfThisSubtask() != 0) {
+                return;
+            }
+            throw new RuntimeException("expected exception.");
+        }
+
+        @Override
+        public void invoke(String value, Context context) throws Exception {
+            NUM_RECEIVED_RECORDS[getRuntimeContext().getIndexOfThisSubtask()]++;
+            assertThat(value).isEqualTo(RECORD);
+        }
+    }
```

Comment:
These classes seem duplications of those in `BlockingShuffleITCase`. We may introduce a `ShuffleITCaseBase` for deduplication.

Reply:
And can `createJobGraph` be deduplicated as well?

Reply:
Abstract common logical to `BatchShuffleITCaseBase`.